File: storm-java21/src/main/java/st/orm/template/ORMTemplate.java
/*
 * Copyright 2024 - 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package st.orm.template;

import java.sql.Connection;
import java.util.List;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;
import javax.sql.DataSource;
import st.orm.Data;
import st.orm.EntityCallback;
import st.orm.StormConfig;
import st.orm.mapping.TemplateDecorator;
import st.orm.repository.EntityRepository;
import st.orm.repository.ProjectionRepository;
import st.orm.repository.RepositoryLookup;
import st.orm.template.impl.ORMTemplateImpl;

/**
 * The primary entry point for Storm's ORM functionality, combining SQL template query construction with
 * repository access.
 *
 * <p>{@code ORMTemplate} extends both {@link QueryTemplate} (for constructing and executing SQL queries) and
 * {@link RepositoryLookup} (for obtaining type-safe {@link EntityRepository} and {@link ProjectionRepository}
 * instances). It is the central interface from which all database operations originate.</p>
 *
 * <p>Instances are created using the static factory methods {@link #of(javax.sql.DataSource)} or
 * {@link #of(java.sql.Connection)}, or via the convenience methods in the {@link Templates} class.
 * For JPA-based usage, see {@code JpaTemplate}.</p>
 *
 * <h2>Example</h2>
 * <pre>{@code
 * ORMTemplate orm = ORMTemplate.of(dataSource);
 *
 * // Repository-based access
 * EntityRepository<User, Integer> users = orm.entity(User.class);
 * Optional<User> user = users.findById(42);
 *
 * // Template-based query
 * List<User> result = orm.query(RAW."""
 *         SELECT \{User.class}
 *         FROM \{User.class}
 *         WHERE \{User_.name} = \{"Alice"}""")
 *     .getResultList(User.class);
 * }</pre>
 *
 * @see Templates
 * @see EntityRepository
 * @see ProjectionRepository
 */
public interface ORMTemplate extends QueryTemplate, RepositoryLookup {

    /**
     * Returns a new {@code ORMTemplate} with the specified entity callback added.
     *
     * <p>The returned template shares the same underlying connection and configuration, but applies the given
     * callback to entity lifecycle operations (insert, update, delete) performed through its repositories. The
     * callback is only invoked for entities matching its type parameter. Multiple callbacks can be registered by
     * chaining calls to this method.</p>
     *
     * @param callback the entity callback to add; must not be {@code null}.
     * @return a new {@code ORMTemplate} with the callback added.
     * @since 1.9
     */
    ORMTemplate withEntityCallback(EntityCallback<?> callback);

    /**
     * Returns a new {@code ORMTemplate} with the specified entity callbacks added.
     *
     * <p>The returned template shares the same underlying connection and configuration, but applies the given
     * callbacks to entity lifecycle operations (insert, update, delete) performed through its repositories. Each
     * callback is only invoked for entities matching its type parameter.</p>
     *
     * @param callbacks the entity callbacks to add; must not be {@code null}.
     * @return a new {@code ORMTemplate} with the callbacks added.
     * @since 1.9
     */
    ORMTemplate withEntityCallbacks(List<EntityCallback<?>> callbacks);

    /**
     * Validates all discovered entity and projection types against the database schema.
     *
     * <p>Logs each validation error and returns the list of error messages. On success, logs a
     * confirmation message and returns an empty list.</p>
     *
     * <p>This method requires a DataSource-backed template. Templates created from a raw
     * {@link Connection} or {@code EntityManager} do not support schema validation.</p>
     *
     * @return the list of validation error messages (empty on success).
     * @throws st.orm.PersistenceException if the template does not support schema validation.
     * @since 1.9
     */
    List<String> validateSchema();

    /**
     * Validates discovered types matching the filter against the database schema.
     *
     * <p>Discovers all entity and projection types via the classpath index, applies the given filter, and validates
     * the matching types. This is useful when a single application connects to multiple datasources and only a
     * subset of types is reachable from each connection.</p>
     *
     * <p>Logs each validation error and returns the list of error messages. On success, logs a
     * confirmation message and returns an empty list.</p>
     *
     * <p>This method requires a DataSource-backed template. Templates created from a raw
     * {@link Connection} or {@code EntityManager} do not support schema validation.</p>
     *
     * @param filter predicate to select which discovered types to validate.
     * @return the list of validation error messages (empty on success).
     * @throws st.orm.PersistenceException if the template does not support schema validation.
     * @since 1.11
     */
    List<String> validateSchema(Predicate<Class<? extends Data>> filter);

    /**
     * Validates the specified types against the database schema.
     *
     * <p>Logs each validation error and returns the list of error messages. On success, logs a
     * confirmation message and returns an empty list.</p>
     *
     * <p>This method requires a DataSource-backed template. Templates created from a raw
     * {@link Connection} or {@code EntityManager} do not support schema validation.</p>
     *
     * @param types the entity and projection types to validate.
     * @return the list of validation error messages (empty on success).
     * @throws st.orm.PersistenceException if the template does not support schema validation.
     * @since 1.9
     */
    List<String> validateSchema(Iterable<Class<? extends Data>> types);

    /**
     * Validates all discovered types and throws if any errors are found.
     *
     * <p>This method requires a DataSource-backed template. Templates created from a raw
     * {@link Connection} or {@code EntityManager} do not support schema validation.</p>
     *
     * @throws st.orm.PersistenceException if validation fails or the template does not support schema validation.
     * @since 1.9
     */
    void validateSchemaOrThrow();

    /**
     * Validates discovered types matching the filter and throws if any errors are found.
     *
     * <p>Discovers all entity and projection types via the classpath index, applies the given filter, and validates
     * the matching types. This is useful when a single application connects to multiple datasources and only a
     * subset of types is reachable from each connection.</p>
     *
     * <p>This method requires a DataSource-backed template. Templates created from a raw
     * {@link Connection} or {@code EntityManager} do not support schema validation.</p>
     *
     * @param filter predicate to select which discovered types to validate.
     * @throws st.orm.PersistenceException if validation fails or the template does not support schema validation.
     * @since 1.11
     */
    void validateSchemaOrThrow(Predicate<Class<? extends Data>> filter);

    /**
     * Validates the specified types and throws if any errors are found.
     *
     * <p>This method requires a DataSource-backed template. Templates created from a raw
     * {@link Connection} or {@code EntityManager} do not support schema validation.</p>
     *
     * @param types the entity and projection types to validate.
     * @throws st.orm.PersistenceException if validation fails or the template does not support schema validation.
     * @since 1.9
     */
    void validateSchemaOrThrow(Iterable<Class<? extends Data>> types);

    /**
     * Returns an {@link ORMTemplate} for use with JDBC.
     *
     * <p>This method creates an ORM repository template using the provided {@link DataSource}.
     * It allows you to perform database operations using JDBC in a type-safe manner.
     *
     * <p>Example usage:
     * <pre>{@code
     * DataSource dataSource = ...;
     * ORMTemplate orm = ORMTemplate.of(dataSource);
     * List<MyTable> otherTables = orm.query(RAW."""
     *         SELECT \{MyTable.class}
     *         FROM \{MyTable.class}
     *         WHERE \{MyTable_.name} = \{"ABC"}""")
     *     .getResultList(MyTable.class);
     * }</pre>
     *
     * @param dataSource the {@link DataSource} to use for database operations; must not be {@code null}.
     * @return an {@link ORMTemplate} configured for use with JDBC.
     */
    static ORMTemplate of(DataSource dataSource) {
        return new ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(dataSource));
    }

    /**
     * Returns an {@link ORMTemplate} for use with JDBC.
     *
     * <p>This method creates an ORM repository template using the provided {@link Connection}.
     * It allows you to perform database operations using JDBC in a type-safe manner.</p>
     *
     * <p><strong>Note:</strong> The caller is responsible for closing the connection after usage.</p>
     *
     * <p>Example usage:
     * <pre>{@code
     * try (Connection connection = ...) {
     *     ORMTemplate orm = ORMTemplate.of(connection);
     *     List<MyTable> otherTables = orm.query(RAW."""
     *             SELECT \{MyTable.class}
     *             FROM \{MyTable.class}
     *             WHERE \{MyTable_.name} = \{"ABC"}""")
     *         .getResultList(MyTable.class)
     * }
     * }</pre>
     *
     * @param connection the {@link Connection} to use for database operations; must not be {@code null}.
     * @return an {@link ORMTemplate} configured for use with JDBC.
     */
    static ORMTemplate of(Connection connection) {
        return new ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(connection));
    }

    /**
     * Returns an {@link ORMTemplate} for use with JDBC, with a custom template decorator.
     *
     * <p>This method creates an ORM repository template using the provided {@link DataSource} and applies
     * the specified decorator to customize template processing behavior.
     *
     * @param dataSource the {@link DataSource} to use for database operations; must not be {@code null}.
     * @param decorator a function that transforms the {@link TemplateDecorator} to customize template processing.
     * @return an {@link ORMTemplate} configured for use with JDBC.
     */
    static ORMTemplate of(DataSource dataSource, UnaryOperator<TemplateDecorator> decorator) {
        return new ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(dataSource, decorator));
    }

    /**
     * Returns an {@link ORMTemplate} for use with JDBC, with a custom template decorator.
     *
     * <p>This method creates an ORM repository template using the provided {@link Connection} and applies
     * the specified decorator to customize template processing behavior.</p>
     *
     * <p><strong>Note:</strong> The caller is responsible for closing the connection after usage.</p>
     *
     * @param connection the {@link Connection} to use for database operations; must not be {@code null}.
     * @param decorator a function that transforms the {@link TemplateDecorator} to customize template processing.
     * @return an {@link ORMTemplate} configured for use with JDBC.
     */
    static ORMTemplate of(Connection connection, UnaryOperator<TemplateDecorator> decorator) {
        return new ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(connection, decorator));
    }

    /**
     * Returns an {@link ORMTemplate} for use with JDBC, configured with the provided {@link StormConfig}.
     *
     * <p>The provided configuration is applied to the template instance, not as a process-wide default.</p>
     *
     * @param dataSource the {@link DataSource} to use for database operations; must not be {@code null}.
     * @param config the Storm configuration to apply; must not be {@code null}.
     * @return an {@link ORMTemplate} configured for use with JDBC.
     */
    static ORMTemplate of(DataSource dataSource, StormConfig config) {
        return new ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(dataSource, config));
    }

    /**
     * Returns an {@link ORMTemplate} for use with JDBC, configured with the provided {@link StormConfig} and a custom
     * template decorator.
     *
     * @param dataSource the {@link DataSource} to use for database operations; must not be {@code null}.
     * @param config the Storm configuration to apply; must not be {@code null}.
     * @param decorator a function that transforms the {@link TemplateDecorator} to customize template processing.
     * @return an {@link ORMTemplate} configured for use with JDBC.
     */
    static ORMTemplate of(DataSource dataSource, StormConfig config,
                          UnaryOperator<TemplateDecorator> decorator) {
        return new ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(dataSource, config, decorator));
    }

    /**
     * Returns an {@link ORMTemplate} for use with JDBC, configured with the provided {@link StormConfig}.
     *
     * <p><strong>Note:</strong> The caller is responsible for closing the connection after usage.</p>
     *
     * @param connection the {@link Connection} to use for database operations; must not be {@code null}.
     * @param config the Storm configuration to apply; must not be {@code null}.
     * @return an {@link ORMTemplate} configured for use with JDBC.
     */
    static ORMTemplate of(Connection connection, StormConfig config) {
        return new ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(connection, config));
    }

    /**
     * Returns an {@link ORMTemplate} for use with JDBC, configured with the provided {@link StormConfig} and a custom
     * template decorator.
     *
     * <p><strong>Note:</strong> The caller is responsible for closing the connection after usage.</p>
     *
     * @param connection the {@link Connection} to use for database operations; must not be {@code null}.
     * @param config the Storm configuration to apply; must not be {@code null}.
     * @param decorator a function that transforms the {@link TemplateDecorator} to customize template processing.
     * @return an {@link ORMTemplate} configured for use with JDBC.
     */
    static ORMTemplate of(Connection connection, StormConfig config,
                          UnaryOperator<TemplateDecorator> decorator) {
        return new ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(connection, config, decorator));
    }

    /**
     * Returns a builder for constructing an {@link ORMTemplate} with instance-scoped integration strategies.
     *
     * <p>The builder is the injection point for platform services: integrations hand their connection provider,
     * transaction template provider, exception mapper and query observer to the template they construct, instead of
     * relying on JVM-global discovery.</p>
     *
     * @param dataSource the {@link DataSource} to use for database operations; must not be {@code null}.
     * @return a builder for constructing the ORM template.
     * @since 1.13
     */
    static Builder builder(DataSource dataSource) {
        return new Builder(st.orm.core.template.ORMTemplate.builder(dataSource));
    }

    /**
     * Returns a builder for constructing an {@link ORMTemplate} backed by a single connection, with instance-scoped
     * integration strategies.
     *
     * <p><strong>Note:</strong> The caller is responsible for closing the connection after usage.</p>
     *
     * @param connection the {@link Connection} to use for database operations; must not be {@code null}.
     * @return a builder for constructing the ORM template.
     * @since 1.13
     */
    static Builder builder(Connection connection) {
        return new Builder(st.orm.core.template.ORMTemplate.builder(connection));
    }

    /**
     * Builder for constructing an {@link ORMTemplate} with instance-scoped integration strategies.
     *
     * @since 1.13
     */
    final class Builder {
        private final st.orm.core.template.ORMTemplate.Builder core;

        private Builder(st.orm.core.template.ORMTemplate.Builder core) {
            this.core = core;
        }

        /**
         * Sets the Storm configuration to apply to the template instance.
         *
         * @param config the Storm configuration; must not be {@code null}.
         * @return this builder.
         */
        public Builder config(StormConfig config) {
            core.config(config);
            return this;
        }

        /**
         * Sets a function that transforms the {@link TemplateDecorator} to customize template processing.
         *
         * @param decorator the decorator function; must not be {@code null}.
         * @return this builder.
         */
        public Builder decorator(UnaryOperator<TemplateDecorator> decorator) {
            core.decorator(decorator);
            return this;
        }

        /**
         * Sets the connection provider used by the template to acquire and release connections.
         *
         * <p>Only valid for data source backed templates; {@link #build()} fails fast otherwise.</p>
         *
         * @param connectionProvider the connection provider; must not be {@code null}.
         * @return this builder.
         */
        public Builder connectionProvider(st.orm.core.spi.ConnectionProvider connectionProvider) {
            core.connectionProvider(connectionProvider);
            return this;
        }

        /**
         * Declares that the data source hands out connections with auto-commit disabled.
         *
         * <p>The declared mode is verified in both directions: a declared template that receives an auto-commit
         * connection fails fast naming the misdeclaration, exactly like an undeclared template that receives a
         * manual-commit one. With the declaration in place, the transactional path performs no auto-commit flips
         * and releases connections in their arrived state; non-transactional connections get auto-commit enabled
         * while Storm uses them and restored before release, so each statement still commits.</p>
         *
         * <p>Cannot be combined with a custom {@link #connectionProvider(st.orm.core.spi.ConnectionProvider)
         * connection provider} and only valid for data source backed templates; {@link #build()} fails fast
         * otherwise.</p>
         *
         * @return this builder.
         * @since 1.14
         */
        public Builder manualCommitConnections() {
            core.manualCommitConnections();
            return this;
        }

        /**
         * Sets the transaction template provider used by the template to participate in transactions.
         *
         * <p>Templates that should share transactions must be configured with the <em>same provider instance</em>.</p>
         *
         * @param transactionTemplateProvider the transaction template provider; must not be {@code null}.
         * @return this builder.
         */
        public Builder transactionTemplateProvider(
                st.orm.core.spi.TransactionTemplateProvider transactionTemplateProvider) {
            core.transactionTemplateProvider(transactionTemplateProvider);
            return this;
        }

        /**
         * Sets the exception mapper that maps failures raised during query execution to the runtime exception thrown
         * to the caller.
         *
         * @param exceptionMapper the exception mapper; must not be {@code null}.
         * @return this builder.
         */
        public Builder exceptionMapper(st.orm.core.spi.ExceptionMapper exceptionMapper) {
            core.exceptionMapper(exceptionMapper);
            return this;
        }

        /**
         * Sets the query observer that is notified of query executions performed by the template.
         *
         * @param queryObserver the query observer; must not be {@code null}.
         * @return this builder.
         */
        public Builder queryObserver(st.orm.core.spi.QueryObserver queryObserver) {
            core.queryObserver(queryObserver);
            return this;
        }

        /**
         * Sets the SQL commenter that appends per-execution comment content to statements, such as the
         * current trace context. Note that per-execution content defeats prepared statement caching.
         *
         * @param sqlCommenter the SQL commenter; must not be {@code null}.
         * @return this builder.
         * @since 1.13
         */
        public Builder sqlCommenter(st.orm.core.spi.SqlCommenter sqlCommenter) {
            core.sqlCommenter(sqlCommenter);
            return this;
        }

        /**
         * Builds the ORM template.
         *
         * @return the ORM template.
         */
        public ORMTemplate build() {
            return new ORMTemplateImpl(core.build());
        }
    }
}

File: storm-kotlin/src/main/kotlin/st/orm/template/ORMTemplate.kt
/*
 * Copyright 2024 - 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package st.orm.template

import st.orm.Data
import st.orm.EntityCallback
import st.orm.StormConfig
import st.orm.mapping.TemplateDecorator
import st.orm.repository.RepositoryLookup
import st.orm.template.impl.ORMTemplateImpl
import java.sql.Connection
import javax.sql.DataSource
import kotlin.reflect.KClass

/**
 * The primary entry point for Storm's ORM functionality in Kotlin, combining SQL template query construction with
 * repository access.
 *
 * `ORMTemplate` extends both [QueryTemplate] (for constructing and executing SQL queries) and
 * [RepositoryLookup] (for obtaining type-safe [st.orm.repository.EntityRepository] and
 * [st.orm.repository.ProjectionRepository] instances). It is the central interface from which all database
 * operations originate.
 *
 * Instances can be created using the companion object factory methods [of], or via the Kotlin extension
 * properties [DataSource.orm] and [Connection.orm]. For JPA-based usage, see `JpaTemplate`.
 *
 * ## Example
 * ```kotlin
 * val orm = dataSource.orm
 *
 * // Repository-based access
 * val users = orm.entity(User::class)
 * val user = users.findById(42)
 *
 * // Template-based query
 * val result = orm.query("SELECT ${User::class} FROM ${User::class} WHERE ${User_.name} = ${"Alice"}")
 *     .getResultList(User::class)
 * ```
 *
 * @see st.orm.repository.EntityRepository
 * @see st.orm.repository.ProjectionRepository
 */
public interface ORMTemplate :
    QueryTemplate,
    RepositoryLookup {

    /**
     * Returns a new [ORMTemplate] with the specified entity callback added.
     *
     * The returned template shares the same underlying connection and configuration, but applies the given
     * callback to entity lifecycle operations (insert, update, delete) performed through its repositories. The
     * callback is only invoked for entities matching its type parameter. Multiple callbacks can be registered by
     * chaining calls to this method.
     *
     * @param callback the entity callback to add.
     * @return a new [ORMTemplate] with the callback added.
     * @since 1.9
     */
    public fun withEntityCallback(callback: EntityCallback<*>): ORMTemplate

    /**
     * Returns a new [ORMTemplate] with the specified entity callbacks added.
     *
     * The returned template shares the same underlying connection and configuration, but applies the given
     * callbacks to entity lifecycle operations (insert, update, delete) performed through its repositories. Each
     * callback is only invoked for entities matching its type parameter.
     *
     * @param callbacks the entity callbacks to add.
     * @return a new [ORMTemplate] with the callbacks added.
     * @since 1.9
     */
    public fun withEntityCallbacks(callbacks: List<EntityCallback<*>>): ORMTemplate

    /**
     * Validates all discovered entity and projection types against the database schema.
     *
     * Logs each validation error and returns the list of error messages. On success, logs a
     * confirmation message and returns an empty list.
     *
     * This method requires a DataSource-backed template. Templates created from a raw
     * [java.sql.Connection] or `EntityManager` do not support schema validation.
     *
     * @return the list of validation error messages (empty on success).
     * @throws st.orm.PersistenceException if the template does not support schema validation.
     * @since 1.9
     */
    public fun validateSchema(): List<String>

    /**
     * Validates discovered types matching the filter against the database schema.
     *
     * Discovers all entity and projection types via the classpath index, applies the given filter, and validates
     * the matching types. This is useful when a single application connects to multiple datasources and only a
     * subset of types is reachable from each connection.
     *
     * Logs each validation error and returns the list of error messages. On success, logs a
     * confirmation message and returns an empty list.
     *
     * This method requires a DataSource-backed template. Templates created from a raw
     * [java.sql.Connection] or `EntityManager` do not support schema validation.
     *
     * @param filter predicate to select which discovered types to validate.
     * @return the list of validation error messages (empty on success).
     * @throws st.orm.PersistenceException if the template does not support schema validation.
     * @since 1.11
     */
    public fun validateSchema(filter: (KClass<out Data>) -> Boolean): List<String>

    /**
     * Validates the specified types against the database schema.
     *
     * Logs each validation error and returns the list of error messages. On success, logs a
     * confirmation message and returns an empty list.
     *
     * This method requires a DataSource-backed template. Templates created from a raw
     * [java.sql.Connection] or `EntityManager` do not support schema validation.
     *
     * @param types the entity and projection types to validate.
     * @return the list of validation error messages (empty on success).
     * @throws st.orm.PersistenceException if the template does not support schema validation.
     * @since 1.9
     */
    public fun validateSchema(vararg types: KClass<out Data>): List<String>

    /**
     * Validates all discovered types and throws if any errors are found.
     *
     * This method requires a DataSource-backed template. Templates created from a raw
     * [java.sql.Connection] or `EntityManager` do not support schema validation.
     *
     * @throws st.orm.PersistenceException if validation fails or the template does not support schema validation.
     * @since 1.9
     */
    public fun validateSchemaOrThrow()

    /**
     * Validates discovered types matching the filter and throws if any errors are found.
     *
     * Discovers all entity and projection types via the classpath index, applies the given filter, and validates
     * the matching types. This is useful when a single application connects to multiple datasources and only a
     * subset of types is reachable from each connection.
     *
     * This method requires a DataSource-backed template. Templates created from a raw
     * [java.sql.Connection] or `EntityManager` do not support schema validation.
     *
     * @param filter predicate to select which discovered types to validate.
     * @throws st.orm.PersistenceException if validation fails or the template does not support schema validation.
     * @since 1.11
     */
    public fun validateSchemaOrThrow(filter: (KClass<out Data>) -> Boolean)

    /**
     * Validates the specified types and throws if any errors are found.
     *
     * This method requires a DataSource-backed template. Templates created from a raw
     * [java.sql.Connection] or `EntityManager` do not support schema validation.
     *
     * @param types the entity and projection types to validate.
     * @throws st.orm.PersistenceException if validation fails or the template does not support schema validation.
     * @since 1.9
     */
    public fun validateSchemaOrThrow(vararg types: KClass<out Data>)

    public companion object {
        /**
         * Returns an [ORMTemplate] for use with JDBC.
         *
         * This method creates an ORM repository template using the provided [DataSource].
         * It allows you to perform database operations using JDBC in a type-safe manner.
         *
         * Example usage:
         * ```
         * DataSource dataSource = ...;
         * ORMTemplate orm = ORMTemplate.of(dataSource);
         * List<MyTable> otherTables = orm.query(RAW."""
         * SELECT \{MyTable.class}
         * FROM \{MyTable.class}
         * WHERE \{MyTable_.name} = \{"ABC"}""")
         * .getResultList(MyTable.class);
         * ```
         *
         * @param dataSource the [DataSource] to use for database operations; must not be `null`.
         * @return an [ORMTemplate] configured for use with JDBC.
         */
        public fun of(dataSource: DataSource): ORMTemplate = ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(dataSource))

        /**
         * Returns an [ORMTemplate] for use with JDBC.
         *
         * This method creates an ORM repository template using the provided [Connection].
         * It allows you to perform database operations using JDBC in a type-safe manner.
         *
         * **Note:** The caller is responsible for closing the connection after usage.
         *
         * Example usage:
         * ```
         * try (Connection connection = ...) {
         * ORMTemplate orm = ORMTemplate.of(connection);
         * List<MyTable> otherTables = orm.query(RAW."""
         * SELECT \{MyTable.class}
         * FROM \{MyTable.class}
         * WHERE \{MyTable_.name} = \{"ABC"}""")
         * .getResultList(MyTable.class)
         * }
         * ```
         *
         * @param connection the [Connection] to use for database operations; must not be `null`.
         * @return an [ORMTemplate] configured for use with JDBC.
         */
        public fun of(connection: Connection): ORMTemplate = ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(connection))

        /**
         * Returns an [ORMTemplate] for use with JDBC, with a custom template decorator.
         *
         * This method creates an ORM repository template using the provided [DataSource] and applies
         * the specified decorator to customize template processing behavior.
         *
         * @param dataSource the [DataSource] to use for database operations.
         * @param decorator a function that transforms the [TemplateDecorator] to customize template processing.
         * @return an [ORMTemplate] configured for use with JDBC.
         */
        public fun of(
            dataSource: DataSource,
            decorator: (TemplateDecorator) -> TemplateDecorator,
        ): ORMTemplate = ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(dataSource, decorator))

        /**
         * Returns an [ORMTemplate] for use with JDBC, with a custom template decorator.
         *
         * This method creates an ORM repository template using the provided [Connection] and applies
         * the specified decorator to customize template processing behavior.
         *
         * **Note:** The caller is responsible for closing the connection after usage.
         *
         * @param connection the [Connection] to use for database operations.
         * @param decorator a function that transforms the [TemplateDecorator] to customize template processing.
         * @return an [ORMTemplate] configured for use with JDBC.
         */
        public fun of(
            connection: Connection,
            decorator: (TemplateDecorator) -> TemplateDecorator,
        ): ORMTemplate = ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(connection, decorator))

        /**
         * Returns an [ORMTemplate] for use with JDBC, configured with the provided [StormConfig].
         *
         * @param dataSource the [DataSource] to use for database operations.
         * @param config the Storm configuration to apply.
         * @return an [ORMTemplate] configured for use with JDBC.
         */
        public fun of(dataSource: DataSource, config: StormConfig): ORMTemplate = ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(dataSource, config))

        /**
         * Returns an [ORMTemplate] for use with JDBC, configured with the provided [StormConfig] and a custom
         * template decorator.
         *
         * @param dataSource the [DataSource] to use for database operations.
         * @param config the Storm configuration to apply.
         * @param decorator a function that transforms the [TemplateDecorator] to customize template processing.
         * @return an [ORMTemplate] configured for use with JDBC.
         */
        public fun of(
            dataSource: DataSource,
            config: StormConfig,
            decorator: (TemplateDecorator) -> TemplateDecorator,
        ): ORMTemplate = ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(dataSource, config, decorator))

        /**
         * Returns an [ORMTemplate] for use with JDBC, configured with the provided [StormConfig].
         *
         * **Note:** The caller is responsible for closing the connection after usage.
         *
         * @param connection the [Connection] to use for database operations.
         * @param config the Storm configuration to apply.
         * @return an [ORMTemplate] configured for use with JDBC.
         */
        public fun of(connection: Connection, config: StormConfig): ORMTemplate = ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(connection, config))

        /**
         * Returns an [ORMTemplate] for use with JDBC, configured with the provided [StormConfig] and a custom
         * template decorator.
         *
         * **Note:** The caller is responsible for closing the connection after usage.
         *
         * @param connection the [Connection] to use for database operations.
         * @param config the Storm configuration to apply.
         * @param decorator a function that transforms the [TemplateDecorator] to customize template processing.
         * @return an [ORMTemplate] configured for use with JDBC.
         */
        public fun of(
            connection: Connection,
            config: StormConfig,
            decorator: (TemplateDecorator) -> TemplateDecorator,
        ): ORMTemplate = ORMTemplateImpl(st.orm.core.template.ORMTemplate.of(connection, config, decorator))

        /**
         * Returns a builder for constructing an [ORMTemplate] with instance-scoped integration strategies.
         *
         * The builder is the injection point for platform services: integrations hand their connection provider,
         * transaction template provider, exception mapper and query observer to the template they construct, instead
         * of relying on JVM-global discovery.
         *
         * @param dataSource the [DataSource] to use for database operations.
         * @return a builder for constructing the ORM template.
         * @since 1.13
         */
        public fun builder(dataSource: DataSource): Builder = Builder(st.orm.core.template.ORMTemplate.builder(dataSource))

        /**
         * Returns a builder for constructing an [ORMTemplate] backed by a single connection, with instance-scoped
         * integration strategies.
         *
         * **Note:** The caller is responsible for closing the connection after usage.
         *
         * @param connection the [Connection] to use for database operations.
         * @return a builder for constructing the ORM template.
         * @since 1.13
         */
        public fun builder(connection: Connection): Builder = Builder(st.orm.core.template.ORMTemplate.builder(connection))
    }

    /**
     * Builder for constructing an [ORMTemplate] with instance-scoped integration strategies.
     *
     * @since 1.13
     */
    public class Builder internal constructor(private val core: st.orm.core.template.ORMTemplate.Builder) {

        /**
         * Sets the Storm configuration to apply to the template instance.
         */
        public fun config(config: StormConfig): Builder = apply { core.config(config) }

        /**
         * Sets a function that transforms the [TemplateDecorator] to customize template processing.
         */
        public fun decorator(decorator: (TemplateDecorator) -> TemplateDecorator): Builder = apply { core.decorator(decorator) }

        /**
         * Sets the connection provider used by the template to acquire and release connections.
         *
         * Only valid for data source backed templates; [build] fails fast otherwise.
         */
        public fun connectionProvider(connectionProvider: st.orm.core.spi.ConnectionProvider): Builder = apply { core.connectionProvider(connectionProvider) }

        /**
         * Sets the transaction template provider used by the template to participate in transactions.
         *
         * Templates that should share transactions must be configured with the *same provider instance*.
         */
        public fun transactionTemplateProvider(transactionTemplateProvider: st.orm.core.spi.TransactionTemplateProvider): Builder = apply { core.transactionTemplateProvider(transactionTemplateProvider) }

        /**
         * Sets the exception mapper that maps failures raised during query execution to the runtime exception thrown
         * to the caller.
         */
        public fun exceptionMapper(exceptionMapper: st.orm.core.spi.ExceptionMapper): Builder = apply { core.exceptionMapper(exceptionMapper) }

        /**
         * Sets the query observer that is notified of query executions performed by the template.
         */
        public fun queryObserver(queryObserver: st.orm.core.spi.QueryObserver): Builder = apply { core.queryObserver(queryObserver) }

        /**
         * Sets the SQL commenter that appends per-execution comment content to statements, such as the
         * current trace context. Note that per-execution content defeats prepared statement caching.
         *
         * @since 1.13
         */
        public fun sqlCommenter(sqlCommenter: st.orm.core.spi.SqlCommenter): Builder = apply { core.sqlCommenter(sqlCommenter) }

        /**
         * Builds the ORM template.
         */
        public fun build(): ORMTemplate = ORMTemplateImpl(core.build())
    }
}

/**
 * Creates an [ORMTemplate] from this [DataSource] with default configuration.
 *
 * ```kotlin
 * val orm = dataSource.orm
 * ```
 *
 * Requires `import st.orm.template.orm`.
 */
public val DataSource.orm: ORMTemplate
    get() = ORMTemplate.of(this)

/**
 * Creates an [ORMTemplate] from this [Connection] with default configuration.
 *
 * Requires `import st.orm.template.orm`.
 */
public val Connection.orm: ORMTemplate
    get() = ORMTemplate.of(this)

/**
 * Creates an [ORMTemplate] from this [DataSource] with a custom [TemplateDecorator].
 *
 * The decorator wraps the template pipeline, allowing cross-cutting concerns such as logging or SQL rewriting.
 *
 * Requires `import st.orm.template.orm`.
 */
public fun DataSource.orm(decorator: (TemplateDecorator) -> TemplateDecorator): ORMTemplate = ORMTemplate.of(this, decorator)

/**
 * Creates an [ORMTemplate] from this [Connection] with a custom [TemplateDecorator].
 *
 * Requires `import st.orm.template.orm`.
 */
public fun Connection.orm(decorator: (TemplateDecorator) -> TemplateDecorator): ORMTemplate = ORMTemplate.of(this, decorator)

/**
 * Creates an [ORMTemplate] from this [DataSource] with a custom [StormConfig].
 *
 * Requires `import st.orm.template.orm`.
 */
public fun DataSource.orm(config: StormConfig): ORMTemplate = ORMTemplate.of(this, config)

/**
 * Creates an [ORMTemplate] from this [Connection] with a custom [StormConfig].
 *
 * Requires `import st.orm.template.orm`.
 */
public fun Connection.orm(config: StormConfig): ORMTemplate = ORMTemplate.of(this, config)

/**
 * Creates an [ORMTemplate] from this [DataSource] with a custom [StormConfig] and [TemplateDecorator].
 *
 * Requires `import st.orm.template.orm`.
 */
public fun DataSource.orm(config: StormConfig, decorator: (TemplateDecorator) -> TemplateDecorator): ORMTemplate = ORMTemplate.of(this, config, decorator)

/**
 * Creates an [ORMTemplate] from this [Connection] with a custom [StormConfig] and [TemplateDecorator].
 *
 * Requires `import st.orm.template.orm`.
 */
public fun Connection.orm(config: StormConfig, decorator: (TemplateDecorator) -> TemplateDecorator): ORMTemplate = ORMTemplate.of(this, config, decorator)
